Token Bridge Feedback (#448)
wormhole-foundation/wormhole-docs: products/token-bridge/guides/transfer-tokens-payload.md

---
title: Transfer Tokens with Payload via the Token Bridge Contract
description: Learn how to transfer your token with an additional payload via Wormhole's Token Bridge for seamless multichain token transfers with payload execution.
categories: Token-Bridge, Transfer
---

# Token Transfers with Payload via the Token Bridge Contract

## Introduction 

> Wormhole's Token Bridge enables seamless cross-chain token transfers using a lock-and-mint mechanism. The bridge locks tokens on the source chain and mints them as wrapped assets on the destination chain. The primary functions of the Token Bridge contracts revolve around:

Reviewer: ```suggestion
Wormhole's [Token Bridge](/docs/products/token-bridge/overview/){target=\_blank} enables seamless cross-chain token transfers using a lock-and-mint mechanism. The bridge locks tokens on the source chain and mints them as wrapped assets on the destination chain. The primary functions of the Token Bridge contracts revolve around:
```

Call `transferTokensWithPayload()` instead of `transferTokens()` to include a custom payload (arbitrary bytes) with the token transfer.

```solidity
function transferTokensWithPayload(
    address token,
    uint256 amount,
    uint16 recipientChain,
    bytes32 recipient,
    uint32 nonce,
    bytes memory payload
) external payable returns (uint64 sequence);
```

??? interface "Parameters"

    `token` ++"address"++
    
    The address of the token being transferred.

    ---

    `amount` ++"uint256"++  
    The amount of tokens to be transferred.

    ---

    `recipientChain` ++"uint16"++  
    The Wormhole chain ID of the destination chain.

    ---

    `recipient` ++"bytes32"++  
    The recipient's address on the destination chain.

    ---

    `nonce` ++"uint32"++  
    A unique identifier for the transaction.

    ---

    `payload` ++"bytes memory"++  
    Arbitrary data payload attached to the transaction.

??? interface "Returns"
    
    `sequence` ++"uint64"++
    
    A unique identifier for the transfer transaction.

> After initiating a transfer on the source chain, the Wormhole Guardian network observes and signs the resulting message, creating a Verifiable Action Approval (VAA). You'll need to fetch this VAA and then call `completeTransferWithPayload()`.

Reviewer: ```suggestion
After initiating a transfer on the source chain, the [Guardian](/docs/protocol/infrastructure/guardians/){target=\_blank} network observes and signs the resulting message, creating a VAA. You'll need to fetch this VAA and then call `completeTransferWithPayload()`.
```

VAA was already mentioned, so no need to spell it out here.

Only the designated recipient contract can redeem tokens. This ensures that the intended contract securely handles the attached payload. On successful redemption, the tokens are minted (if foreign) or released (if native) to the recipient address on the destination chain. For payload transfers, the designated contract can execute the payload's logic at this time.

```solidity

> function completeTransferWithPayload(bytes memory encodedVm) external returns (bytes memory);

Reviewer: ```suggestion
function completeTransferWithPayload(
    bytes memory encodedVm
) external returns (bytes memory);
```

```

??? interface "Parameters"

    `encodedVm` ++"bytes memory"++

    The signed VAA containing the transfer details.

??? interface "Returns"

>     `bytes memory`

Reviewer: ```suggestion
    ++"bytes memory"++
```
